### ISSUE TITLE
Healthchecks integration

### ISSUE THREAD

**Issue by Zottelchen** [creativeprojects/resticprofile#102]
Hi,
I have a feature request: An Integration of Healthchecks. There is a [hosted version](https://healthchecks.io/), but it can also be [selfhosted](https://github.com/healthchecks/healthchecks). It should do the following things:

1. Webhook when starting a location
2. Webhook when finishing a location, including status (or success/fail) as well as a (truncated?) log.

This would be similar to e.g., [borgmatic](https://torsion.org/borgmatic/docs/how-to/monitor-your-backups/#healthchecks-hook). The benefit of having that directly integrated here is that it is one thing less to break. Or in a Windows case, one thing less to properly install [curl] as well as not having to fight with questionable Windows batch syntax just to set up a backup.

**Comment by creativeprojects** [comment]
It shouldn't be difficult if we can reuse the framework used by the status file and prometheus (and at first glance it looks possible)

**Comment by bdillahu** [comment]
Right now I’m just using the run_before, etc. to call the health checks start/success/fail pings, but having something built in would be slick

**Comment by creativeprojects** [comment]
So here's the proposal:

It's adding 3 flags. 2 flags on the profile level:
- `healthchecks-url`
- `healthchecks-timeout`

and 1 flag at the command level:
- `healthchecks-uuid`

Even though it's called `healthchecks-uuid` it should actually work with a slug as well.

All the schedulable commands should accept the `healthchecks-uuid` field (backup, check, forget, prune and copy).

It should send a "start" before the command start and a "success" of "fail" at the end. Apparently the service cannot be configured to **not** fail on exit code 3 (restic warning) so you have to decide if it's a fail of a success with the usual `no-error-on-warning` field.

If the service cannot be contacted before the timeout (10 seconds by default) it will display an error message but will **not** fail the run.

Does that sound all-right?

**Comment by bdillahu** [comment]
Instead of a separate URL and UUID field, I wonder about just having a URL - then people could potentially use it with other, similar services that might use a different structure.

Alternatively, now that I’ve used this longer with run_before taking the place, is that really the best solution - just use a run_before/after if you want something like this and not complicate the code more???

Either way I’m good with, and your proposal would certainly seem to work fine.

**Comment by creativeprojects** [comment]
Without knowing all the possible systems that people would want to use, you can never come up with anything *generic* that actually works.

There's the http method (head, get, post) there's the authentication (API key, basic auth, etc.) and the body if you send a post. Healthchecks accepts some plain text in a POST but we can be sure some other system would only accept some JSON format.

I'm also not sure if it's a good idea to either hardcode it for each provider, or to try to make it generic (and probably fail)

**Comment by bdillahu** [comment]
Good point… simple and works might just be for the best. If other popular things come along, either add or reconsider once you can compare the api.

you have my vote :-)

**Comment by Zottelchen** [comment]
Sounds good to me. Will it also send the log?

**Comment by creativeprojects** [comment]
Right, so I've completely changed my mind! Don't worry it happens all the time 😄 

How about something actually generic? I've looked at the different solutions out there and it looks like it would work for any provider using this kind of configuration:

```yaml
{{ $healthcheckUUID := "831e288e-1293-46f8-ac31-70ea7f875650" }}

profile:
    repository: "/Volumes/RAMDisk/{{ .Profile.Name }}"
    password-file: ../key
    send-before:
        method: HEAD
        url: https://hc-ping.com/{{ $healthcheckUUID }}/start
        headers:
          - name: User-Agent
            value: "example of header value"

    send-after:
        # send 2 of these for fun
        - method: POST
          url: https://hc-ping.com/{{ $healthcheckUUID }}
          body: "${ERROR_STDERR}"
        - method: HEAD
          url: https://hc-ping.com/{{ $healthcheckUUID }}

    send-after-fail:
        method: POST
        url: https://hc-ping.com/{{ $healthcheckUUID }}/fail
        body: "${ERROR}\n\n${ERROR_STDERR}"

    backup:
        no-error-on-warning: true
        source: {{ .CurrentDir }}
        exclude:
          - "/**/.git/"

```


For the simplicity of the main configuration, it can be made into a template and loaded from an include file

> Sounds good to me. Will it also send the log?

Do you mean the full logs, or just the error logs?

**Comment by bdillahu** [comment]
It looks like for healthchecks.io at least, they recommend sending the last 100KB of the log.

See: https://healthchecks.io/docs/attaching_logs/

I would just send logs of error stuff myself - I self-host the healthchecks and don't need another copy of the main logs. The error log might be handy to just see it right there without going and digging it out, but personally I haven't done any logging with healthchecks.

**Comment by Zottelchen** [comment]
Originally thought about the full log, but yeah the error log is probably enough 👍

**Comment by creativeprojects** [comment]
This is going to be released with v0.18.0
In the meantime you can install resticprofile from source if you want to give it a try 👍🏻 

Please note the configuration changed a bit since these posts, the correct configuration is described in the Pull Request description, or in the documentation when ready https://creativeprojects.github.io/resticprofile/configuration/http_hooks/
